unittests/finality_tests.cpp

```diff
@@ -0,0 +1,739 @@
+   }
+
+   // returns true if node3's LIB has advanced
+   bool node3_lib_advancing() {
+      return lib_advancing(node3.control->if_lib_num(), node3_prev_lib_bum);
+   }
+
+   // Produces a number of blocks and returns true if LIB is advancing.
+   // This function can be only used at the end of a test as it clears
+   // bob_votes and carol_votes when starting.
+   bool produce_blocks_and_verify_lib_advancing() {
+      // start from fresh
+      bob_votes.clear();
+      carol_votes.clear();
+
+      for (auto i = 0; i < 30; ++i) {
+         produce_and_push_block();
+         process_bob_vote();
+         if (!node1_lib_advancing() || !node2_lib_advancing() || !node3_lib_advancing()) {
+            return false;
+         }
+      }
+
+      return true;
+   }
+
+   std::vector<eosio::chain::vote_message> bob_votes;
+
+private:
+   bool lib_advancing(uint32_t curr_lib_num, uint32_t& prev_lib_num) {
+      auto advancing = curr_lib_num > prev_lib_num;
+      // update pre_lib_num for next time check
+      prev_lib_num = curr_lib_num;
+      return advancing;
+   }
+
+   eosio::testing::tester   node1;
+   eosio::testing::tester   node2;
+   eosio::testing::tester   node3;
+   uint32_t                 node1_prev_lib_bum {0};
+   uint32_t                 node2_prev_lib_bum {0};
+   uint32_t                 node3_prev_lib_bum {0};
+   std::vector<eosio::chain::vote_message> carol_votes;
+
+   void setup_node(eosio::testing::tester& node, eosio::chain::account_name local_finalizer) {
+      std::vector<eosio::chain::account_name> account_names = { "alice"_n, "bob"_n, "carol"_n };
+
+      node.produce_block();
+      node.create_accounts(account_names);
+      node.produce_block();
+
+      // activate hotstuff
+      eosio::testing::base_tester::finalizer_policy_input policy_input = {
+         .finalizers       = { {.name = "alice"_n, .weight = 1},
+                               {.name = "bob"_n, .weight = 1},
+                               {.name = "carol"_n, .weight = 1}},
+         .threshold        = 2,
+         .local_finalizers = {local_finalizer}
+      };
+      node.set_finalizers(policy_input);
+      auto block = node.produce_block();
+
+      // this block contains the header extension for the instant finality
+      std::optional<eosio::chain::block_header_extension> ext = block->extract_header_extension(eosio::chain::instant_finality_extension::extension_id());
+      BOOST_TEST(!!ext);
+      std::optional<eosio::chain::finalizer_policy> fin_policy = std::get<eosio::chain::instant_finality_extension>(*ext).new_finalizer_policy;
+      BOOST_TEST(!!fin_policy);
+      BOOST_TEST(fin_policy->finalizers.size() == account_names.size());
+      BOOST_TEST(fin_policy->generation == 1);
+   }
+};
+
+// verify LIB advances with 2 finalizers voting and for a long time
+BOOST_AUTO_TEST_CASE(two_votes_large_number_blocks) { try {
+   tester_cluster cluster;
+
+   for (auto i = 0; i < 1000; ++i) {
```

Comment:
Seems pointless to loop so many times. How could anything be different after say the 2nd loop?

Reply:
I just don't see how it could make any difference how long the test runs.

Reply:
Was thinking about testing a long duration of block production. Here is probably not a right place to do it. After reducing all to 3, running time of all tests in this suite is reduced from 56 seconds to 12 seconds.